
<center><h1>TP3 : Débuter sur Python



# I. Premiers exemples

On considère le programme Python suivant :

In [ ]:
a=5
b=a+3
c=2*b-a

1. Que vaut la variable `c` à la fin de ce programme ?




> Réponse : 11

2. Pour vérifier ce résultat, selectionner le programme suivant et cliquer sur le bouton "Run" dans la barre d'outils en haut de la page (ou entrer le raccourci Ctrl+Entrer):

In [1]:
a=5
b=a+3
c=2*b-a
print(c)

11


3. Aviez-vous trouvé le bon résultat ?

> Réponse : Oui



4. À quoi sert la commande `print(c)` ? 

> Réponse : À afficher la valeur de la variable `c`

On considère le programme Python suivant :

In [ ]:
x=7
x=x+2
y=x*(x-2)
z=y+x

5. Que vaut la variable `z` à la fin de ce programme ?




> Réponse : 72



6. Modifier le programme dans le champ ci-dessous afin d'afficher en sortie la valeur de `z` et ainsi vérifier votre résultat :

In [2]:
x=7
x=x+2
y=x*(x-2)
z=y+x
print(z)

72


7. Aviez-vous trouvé le bon résultat ?

> Réponse : Oui



8. L'égalité `x=x+2` vue dans le programme précédent est-elle vraie en général en mathématiques ?

> Réponse : Non



**En langage Python, le `=` permet de faire l'affectation d'une variable.**

**L'affectation d'une variable est l'instruction permettant de "remplir" le *casier* correspondant.**



9. Tester le programme Python suivant :

In [3]:
x=int(input('Choisis un nombre : '))
y=x+2
z=y+5
print('On obtient',z)

Choisis un nombre : 28
On obtient 35


10. À quoi sert la commande `input` ?



> Réponse : À demander à l'utilisateur de rentrer des données

11. Dans le champ ci-dessous, créer un programme qui :



* Demande un nombre à l'utilisateur
* Ajoute 7 à ce nombre
* Met au carré le résultat obtenu
* Ajoute au résultat le nombre de départ
* Divise le résultat par 2
* Affiche le résultat obtenu

In [1]:
x=int(input('Choisis un nombre : '))
y=x+7
z=y**2
w=z+x
t=w/2
print('On obtient',t)

Choisis un nombre : 28
On obtient 626.5


On peut remarquer que le format du résultat affiché à l'issue de ce programme est légèrement différent de celui du programme précédent. Cela s'explique par l'existence en Python de plusieurs types de variables, que nous allons découvrir dans la partie suivante. 

# II. Les différents types de variables



1. Lancer les programmes ci-dessous :

In [2]:
a=2
type(a)

int

In [3]:
b=428
type(b)

int

In [4]:
c=2.5
type(c)

float

In [5]:
d=10.456
type(d)

float

In [6]:
e=15/2
type(e)

float

In [7]:
f='bonjour'
type(f)

str

In [8]:
g='IloveSNT'
type(g)

str

In [9]:
h=[1,2,3]
type(h)

list

In [10]:
i=[10,15,20,25,30]
type(i)

list

2. Quel est le type de chaque variable ? 

> Réponse : `a` et `b` sont des nombres entiers

> `c`, `d` et `e` sont des nombres réels

> `f` et `g` sont des chaînes de caractères

> `h` et `i` sont des listes 

**Une variable contient une valeur. On peut rencontrer plusieurs types de valeurs :**

- **`int` : les nombres entiers** 
- **`float` : les nombres réels** 
- **`str` : les chaînes de caractères (*string* en anglais), entre deux apostrophes `'texte'`** 
- **`list` : les listes d'éléments, entre deux crochets `[ ]`**

**Bien entendu, on ne pourra pas réaliser certaines opérations entre deux variables qui ne sont pas du même type, d'où l'importance de bien savoir le type de chaque variable de votre programme.**

On considère le programme Python suivant : 

In [11]:
a=int(input())
b=float(input())
a=a*b
a=a+2*b
a=a**2
print(a)

2
1.5
36.0


3. Si l'utilisateur rentre les valeurs 2 pour la variable `a` et 1,5 pour la variable `b`, que va afficher le programme une fois son exécution terminée ?

> Réponse : 36

4. À quoi correspond la double étoile `**` dans le programme ?

> Réponse : cela signifie "puissance"

On peut remarquer qu'il n'y a pas  de problème pour utiliser les opérations traditionnelles (addition, multiplication, division..) entre des variables de type entier et réel. 

Nous reviendrons dans un autre TP sur les opérations sur les variables de type `list` ou `str`.

# III. Les fonctions

On considère la fonction Python suivante :

In [ ]:
def aire(c):
    res=c**2
    return res

1. Qu'affichera la commande `print(aire(5))` ?

> Réponse : 25

2. Lancer le programme suivant afin de vérifier votre résultat :

In [12]:
def aire(c):
    res=c**2
    return res

print(aire(5))

25


3. Que permet de calculer cette fonction ?

> Réponse : l'aire d'un carré de côté `c`

3. Compléter la fonction suivante afin qu'elle renvoie le périmètre d'un carré de côté `c` donné en paramètre :

In [13]:
def perimetre(c):
    res=c*4
    return res

print(perimetre(10))

40


4. Ajouter l'instruction `print(perimetre(10))` à la suite de votre fonction afin de la tester.

On considère la fonction Python suivante :

In [ ]:
def fonction(a,b):
    p=a+b
    r=2*a*b
    s=p-r
    return s

5. Sans l'éxecuter, qu'affichera la commande `print(fonction(3,2)` ?

> Réponse : -7

6. Sans l'éxecuter, qu'affichera la commande `print(fonction(5,8)` ?

> Réponse : -67

7. Utiliser le programme suivant afin de vérifier vos résultats :

In [14]:
def fonction(a,b):
    p=a+b
    r=2*a*b
    s=p-r
    return s

print(fonction(5,8))

-67


**Une *fonction* est une succession d'instructions regroupées dans un bloc que l'on désigne par un nom. En général, une fonction a une entrée (des informations qu'on lui fournit appelées *arguments* ou *paramètres*), exécute un travail sur ces arguments et renvoie des informations en sortie, comme des nombres et/ou des messages.**

**On définit une telle fonction en Python par :**

In [ ]:
def nom_de_la_fonction(argument1,argument2,...):
  instructions éventuelles 
  ...
  return résultat1,résultat2,... 


**Utiliser des fonctions permet d'effectuer des opérations répétitives sans devoir saisir les instructions à chaque fois. Cela permet également de découper un programme en parties plus simples et de rendre le programme plus lisible.**

Par exemple, dans le programme ci-dessous, la fonction `div_eucl` prend **deux arguments** et renvoit **deux résultats**.

In [15]:
def div_eucl(a,b):
    q=int(a/b)
    r=a%b
    return q,r

div_eucl(51,4)

(12, 3)

8. Quelles seront les résultats retournés par l'instruction `div_eucl(51,4)` ?

> Réponse : 12, 3

9. Tester l'instruction dans le champ contenant la fonction pour tester votre résultat. 

10. Dans l'encadré ci-dessous, écrire une fonction `puissance` qui prend deux arguments `x` et `n` et qui retourne la valeur de `x` à la puissance `n`.

In [ ]:
def puissance(x,n):
    return x**n

# IV. Le module math

1. Lancer le programme ci-dessous :

In [16]:
a=sqrt(4)
print(a)

NameError: name 'sqrt' is not defined

On remarque que le programme ne fonctionne pas. Certains outils de Python sont rangés dans des modules (ou bibliothèques). Quand on veut utiliser un tel outil, on doit l'importer depuis le module. On peut importer l'outil seul ou l'ensemble des outils du module. En rajoutant, `from math import sqrt` au début de notre programme précédent comme ci-dessous, le programme fonctionne. 

In [6]:
from math import sqrt

a=sqrt(4)
print(a)

2.0


2. À quoi correspond l'outil `sqrt` ? (Si besoin, chercher sur Internet la signification de `sqrt`)

> Réponse : cela correspond à "racine"

Si notre programme comporte plusieurs outils issus du module math, au lieu de les importer un par un, on peut directement tous les importer en tapant `from math import *`. Par exemple :


In [ ]:
from math import *

def vol_piscine_hexagonale(cote,hauteur):
    vol=hauteur*3*sqrt(3)*cote/2
    return vol

def vol_piscine_ronde(rayon,hauteur):
    vol=hauteur*pi*rayon**2

3. Dans le programme ci-dessus, quel autre outil que `sqrt` a nécessité l'importation du module `math` ?

> Réponse : pi

Dans le module `math`, toutes les fonctions mathématiques sont présentes (cosinus, sinus, arcsinus, exponentielle, pi, racine carrée...). Cependant, il existe d'autres modules comme par exemple `numpy` pour fabriquer des tableaux ou encore `matplot` pour faire des graphiques ou des dessins. 

4. Dans le champ ci-dessous, écrire une fonction qui prend comme paramètre les longueurs des plus petits côtés d'un triangle rectangle et qui retourne la longueur de l'hypoténuse.

In [7]:
from math import *

def hypotenuse(a,b):
    res=sqrt(a**2+b**2)
    return res

# V. Les instructions conditionnelles *Si ... Alors ... Sinon*

1. Qu'affiche le programme suivant ? Expliquer ce résultat.

In [8]:
x=int(input('Entrer un nombre'))
if x<=50 :
  print('Trop petit')
elif x>=100 :
  print('Trop grand')
else :
  print('Convient')

Entrer un nombre28
Trop petit


> Réponse : le programme affiche "Trop petit" quand on rentre un nombre plus petit que 50, et "Trop grand" quand on rentre un nombre plus grand que 100.

2. Compléter le programme suivant :

In [ ]:
age=int(input('Entrez votre âge'))
if age>=18 :
  print('Vous êtes majeur')
else :
  print('Vous êtes mineur')

3. Créer un programme qui prend en paramètres deux nombres `a` et `b` et qui affiche si leur somme est plus grande ou plus petite que 100 :

In [ ]:
a=float(input())
b=float(input())
if a+b>=100:
    print('Plus grand')
else :
    print('Plus petit')

**Dans un algorithme, il peut être utile d'éxecuter un bloc d'instructions uniquement lorsque certaines conditions sont satisfaites. On utilise alors la strucure conditionnelle *Si ... Alors ... Sinon*, traduite en langage Python par les commandes `if ... :`, `else :`**

**Lorsqu'on les structures conditionnelles, on peut comparer des valeurs en utilisant les symboles suivants :**

<img src="https://drive.google.com/uc?export=download&id=1NST2U81oAkEofo1GScVeUI2nytEHgoQH"/>  



Attention, il faut utiliser `==` et non pas `=` lorsqu'on souhaite vérifier une égalité (le `=` étant utilisé pour l'affectation des variables).

4. Lancer le programme ci-dessous :  

In [24]:
#Programme jeu de hasard

from random import *

x=randint(0,9)

rep=int(input('Choisissez un chiffre (entre 0 et 9)'))
if rep==x :
    print('Vous avez gagné !!!')
else :
    print('Vous avez perdu...')

Choisissez un chiffre (entre 0 et 9)8
Vous avez perdu...


5. À quoi sert la commande `randint(0,9)` ?

> Réponse : À tirer un nombre entier entre 0 et 9 aléatoirement

*Nous essaierons d'améliorer ce programme de jeu dans la suite de ce TP*

6. Quelle est l'utilité de la fonction `pyth` du programme suivant ?

In [20]:
from math import sqrt

def pyth(cote1,cote2):
    rep=input("Vous désirez calculer la longueur de l'hypoténuse ?")
    if rep=='oui':
        longueur=sqrt(cote1**2+cote2**2)
    else :
        if cote1<cote2:
            longueur=sqrt(cote2**2-cote1**2)
        else :
            longueur=sqrt(cote1**2-cote2**2)
    return longueur

print(pyth(10,8))

Vous désirez calculer la longueur de l'hypoténuse ?non
6.0


> Réponse : Elle sert à calculer la longueur du troisième côté dans un triangle rectangle.

7. Sans la lancer, qu'affichera la commande `print(pyth(10,8))` si on répond `non` ?

> Réponse : 6

8. Lancer cette commande à la suite du programme ci-dessus pour vérifier votre réponse.

**Il peut y avoir plusieurs conditions à vérifier pour réaliser un bloc et on peut avoir plus de deux blocs conditionnels comme dans l'exemple ci-dessous :**

In [22]:
from random import random

x=randint(0,9)
y=randint(0,9)
rep=int(input('Choisissez un chiffre (entre 0 et 9)'))

if x==rep and y<=rep :
    print('GAGNE !!!')
if x==rep or y>rep :
    print('GAGNE !!!')
else :
    print('PERDU...')

Choisissez un chiffre (entre 0 et 9)8
PERDU...


9. Tester le programme ci-dessus jusqu'à gagner. Préciser le nombre d'essai :

> Réponse : 3

10. On reprend le programme *jeu de hasard* vu précédemment. Rajouter dans le bloc `else` une structure conditionnelle qui permet au joueur de savoir, lorsqu'il a perdu, si le nombre qu'il avait choisi était plus grand ou plus petit que le nombre qu'il fallait trouver :

In [25]:
#Programme jeu de hasard

from random import *

x=randint(0,9)

rep=int(input('Choisissez un chiffre (entre 0 et 9)'))
if rep==x :
    print('Vous avez gagné !!!')
else :
    print('Vous avez perdu...')
    if rep>x :
         print("C'est trop grand")
    else :
         print("C'est trop petit")

Choisissez un chiffre (entre 0 et 9)8
Vous avez perdu...
C'est trop grand


# VI. Les boucles *for*

1. Lancer le programme ci-dessous. Expliquer comment le résultat a été obtenu.

In [26]:
x=0
for i in range(1,11):
    x=x+4
    print(x)

4
8
12
16
20
24
28
32
36
40


> Réponse : l'instruction x=x+4 s'est répétée 10 fois

**Pour répéter des instructions un nombre de fois connu, on utilise une boucle `for` (ou boucle *Pour*).**

**`range(1,N+1)` est la liste des entiers consécutifs de 1 à N inclus. Ainsi, `range(1,11)` est la liste des entiers de 1 à 10 inclus.**

2. Sans le lancer, déterminer ce que va afficher ce programme :

In [27]:
x=1
for i in range(1,5):
    x=x*3
    print(x)

3
9
27
81


> Réponse : 3
9
27
81

3. Lancer le programme pour vérifier le résultat.

4. Sans le lancer, déterminer ce que va afficher ce programme :

In [28]:
x=2
for i in range(1,6):
    x=x*i
    print(x)

2
4
12
48
240


> Réponse : 2
4
12
48
240

5. Lancer le programme pour vérifier le résultat.

6. Une fourmilière compte 200 fourmis. Chaque année, cette population augmente de 5\%. On souhaite étudier son évolution sur 10 ans. Pour cela, compléter le programme suivant afin qu'il affiche le nombre de fourmis dans 10 ans :

In [29]:
pop=200
for i in range(1,11) :
    augmentation=pop*5/100
    pop=pop+augmentation
print(pop)

325.7789253554882


7. Lancer les programmes suivants :

In [30]:
a=3
for i in range(1,5):
    a=a+4
    a=a+5
print(a)

39


In [31]:
a=3
for i in range(1,5):
    a=a+4
a=a+5
print(a)

24


Comment expliquer la différence entre les résultats :

> Réponse : dans le premier, l'instruction a=a+5 a été effectuée à chaque itération de la boucle, alors qu'elle n'a été effectuée qu'une fois, à la fin, dans le deuxième


**En Python, *l'indentation automatique* des instructions est très importante. L'annulation de l'incrémentation marque la sortie de la structure.**

8. Si nous reprenions le jeu de hasard de la partie précédente et que nous le rendions plus intéressant :  recopier le programme de la question 10 de la partie précédente et créer une boucle qui permettrait à l'utilisateur de tenter 5 fois sa chance sans avoir à relancer le programme.

# VII. Les boucles *while*

1. Lancer le programme ci-dessous. Expliquer comment le résultat a été obtenu.

In [ ]:
x=100
n=0
while x>50:
    x=x-6
    n=n+1
print(n)

> Réponse :

**Lorsque le nombre de répétitions n'est à priori pas connu à l'avance, mais que l'on connait une condition d'arrêt, on utilise une boucle `while` (ou boucle *Tant que*).**

2. Sans le lancer, déterminer ce que va afficher ce programme :

In [ ]:
a=1
while a<30:
  a=a*4
print(a)

> Réponse : 

3. Lancer le programme pour vérifier le résultat.

4. Sans le lancer, déterminer ce que va afficher ce programme :

In [ ]:
n=0
while 2**n<1000:
  print(2**n)
  n=n+1

> Réponse : 

5. Lancer le programme pour vérifier le résultat.

**Ici, la variable `n` sert de *compteur* : on l'initialise à 0 au début et on l'incrémente à chaque passage dans la boucle.**

6. En reprenant le programme vu à la question 6. de la partie précédente, créer un programme qui affiche le nombre d'années nécessaires pour que la fourmilière dépasse 1000 fourmis.

7. Reprendre le jeu de hasard de la partie précédente : modifier le programme afin de permettre au joueur de faire autant d'essais que nécessaire pour gagner et d'afficher ce nombre. 

# VII. Si tu as de l'avance...

1. Que permet de faire la fonction suivante ?

In [ ]:
def carre():
  for i in range(1,6):
    for j in range(1,6):
       print('*', end=' ')
    print()

carre()

> Réponse :

2. Créer une fonction `carre(n)` qui dessine de la même manière un carré de taille `n`.

2. Créer une fonction `rectangle(largeur,longueur)`qui dessine un rectangle pour lequel les dimensions sont données en paramètres.

Par exemple, `rectangle(5,8)` doit afficher :

<img src="https://drive.google.com/uc?export=download&id=1psco6tEAYXcCGmMBrfuNp9TzMQB1Tnnn" width="130"/>

3. Créer une fonction `diagonale(n)`, avec laquelle `diagonale(6)` doit afficher :

<img src="https://drive.google.com/uc?export=download&id=1-_kBLRoUuLFVYGkxA5VfXWNyT1zSfYRJ" width="130"/>

4. Créer une fonction `triangle(n)`, avec laquelle `triangle(6)` doit afficher :

<img src="https://drive.google.com/uc?export=download&id=10Aq2RqfJ-8CECc4bsUy4mR_3XPeYlck5" width="130"/>

5. Créer une fonction `triangle2(n)`, avec laquelle `triangle2(6)` doit afficher :

<img src="https://drive.google.com/uc?export=download&id=1NLpHhSWjIF06bXR0RaKTTB2AjjewtNSk" width="130"/>

5. Créer une fonction `triangle3(n)`, avec laquelle `triangle3(6)` doit afficher :

<img src="https://drive.google.com/uc?export=download&id=1g3MSl0xeeSMUfLtddHY6QKszemNJACyN" width="130"/>

6. Créer une fonction `triangle4(n)`, avec laquelle `triangle4(6)` doit afficher :

<img src="https://drive.google.com/uc?export=download&id=1FIqiTM3MlmPGb2lBNtuBK_nKxXBf2svu" width="130"/>

7. Créer une fonction `sapin(n)`, avec laquelle `sapin(6)` doit afficher :

<img src="https://drive.google.com/uc?export=download&id=1eQwLlNfmarLA-xDxYWafwC_kbq0lchXY" width="130"/>

8. Créer une fonction `sapin2(n)` qui crée des sapins plus élégants, avec laquelle `sapin2(6)` doit afficher :

<img src="https://drive.google.com/uc?export=download&id=162ZRmiVamoKWABQqQxUvPf1TshmukD7M" width="250"/>